doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,148 @@
+.. _custom_datasources:
+
+==================
+Custom Datasources
+==================
+
+Ray Datasets supports multiple ways to :ref:`create a dataset <creating_datasets>`, 
+allowing you to easily ingest data of common formats from popular sources. However, if the 
+datasource you want to read from is not in the built-in list, don't worry, you can implement 
+a custom one for your use case. In this guide, we will walk you through how to build 
+your own custom datasource, using MongoDB as an example. By the end of the guide, you will have a datasource
+that you can use as follows:
+
+.. code-block:: python
+
+    # Read from custom MongoDB datasource.
+    # The args are passed to MongoDatasource.create_reader().
+    ds = ray.data.read_datasource(
+        MongoDatasource(),
+        uri=MY_URI, database=MY_DATABASE,
+        collection=MY_COLLECTION, pipelines=MY_QUERIES
+    )
+
+    # Write to custom MongoDB datasource.
+    # The args are passed to MongoDatasource.do_write().
+    ds.write_datasource(
+        MongoDatasource(), uri=MY_URI, database=MY_DATABASE, collection=MY_COLLECTION
+    )
+
+
+A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the 
+example here, let's call it ``MongoDatasource``. At a high level, it will have two 
+core parts to build out:
+
+* read support with :meth:`create_reader() <ray.data.Datasource.create_reader>`
+* write support with :meth:`do_write() <ray.data.Datasource.do_write>`.
+
+Here are the key design choices we will make in this guide:
+
+-  **MongoDB connector**: We use `PyMongo <https://pymongo.readthedocs.io/en/stable/>`__ to connect to MongoDB.
+-  **MongoDB to Arrow conversion**: We use `PyMongoArrow <https://mongo-arrow.readthedocs.io/en/latest/>`__ to convert query results into Arrow tables, which Datasets supports as a block format.
+-  **Parallel execution**: We ask the user to provide a list of MongoDB queries, with each corresponding to a read task (i.e. a :class:`~ray.data.ReadTask`) that Datasets will execute in parallel.
+
+------------
+Read support
+------------
+
+To support reading, we implement :meth:`create_reader() <ray.data.Datasource.create_reader>`, returning a :class:`~ray.data.Datasource.Reader` implementation for 
+MongoDB. This ``Reader`` creates a list of :class:`~ray.data.ReadTask` for the given 
+list of MongoDB queries. Each :class:`~ray.data.ReadTask` returns a list of blocks when called, and
+each :class:`~ray.data.ReadTask` is executed in remote workers to parallelize the execution.
+
+You can find documentation about :ref:`Ray Data blocks here <block-api>`.
```

Comment:
Branding
```suggestion
You can find documentation about :ref:`Ray Datasets blocks here <block-api>`.
```

Reply:
Also, we might want to link out to the Datasets data model Key Concepts section to give a better idea of what these "blocks" are. https://docs.ray.io/en/master/data/key-concepts.html#datasets